### ISSUE TITLE
MySQL database empty when using start (and start-dev) with minkube and db hosted on localhost

### ISSUE THREAD

**Issue by enfochannel** [keycloak/keycloak#14999]
### Describe the bug

Similar to reference #10959 I am unable to see any data persisted in local MySQL DB.

minikube version: v1.24.0
mysql  Ver 8.0.31 for macos12.6 on x86_64 (Homebrew)

deployment and service:
```
apiVersion: v1
kind: Service
metadata:
  name: keycloak
  labels:
    app: keycloak
spec:
  ports:
  - name: http
    port: 8080
    targetPort: 8080
  selector:
    app: keycloak
  type: LoadBalancer
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: keycloak
  labels:
    app: keycloak
spec:
  replicas: 1
  selector:
    matchLabels:
      app: keycloak
  template:
    metadata:
      labels:
        app: keycloak
    spec:
      containers:
      - name: keycloak
        image: quay.io/keycloak/keycloak:19.0.2
        args: ["start"]
        env:
        - name: KEYCLOAK_ADMIN
          value: "admin"
        - name: KEYCLOAK_ADMIN_PASSWORD
          value: "password"
        - name: KC_PROXY
          value: "edge"
        - name: DB_VENDOR
          value: mysql
        - name: DB_ADDR
          value: "192.168.64.1" #(host.minikube.internal) the IP to connect to localhost db from container.
        - name: DB_PORT
          value: "3306"
        - name: DB_DATABASE
          value: "keycloak"
        - name: DB_USER
          value: "root"
        - name: DB_PASSWORD
          value: "rootpassword"
        - name: JDBC_PARAMS
          value: "serverTimezone=UTC" #DB timezone is different and thus this extra parameter was required.
        - name: KC_HOSTNAME
          value: keycloak.192.168.64.19.nip.io
        ports:
        - name: http
          containerPort: 8080
        readinessProbe:
          httpGet:
            path: /realms/master
            port: 8080
```
Ingress service running as is. I can access the admin using above ingress host (KC_HOSTNAME)

kubectl get pods -w
NAME                       READY   STATUS              RESTARTS   AGE
keycloak-c5c7b8877-gdlfm   1/1     Running             0          2m11s

No logs were reported.

/opt/keycloak/bin/kc.sh show-config

2022-10-18 21:59:19,968 WARN  [io.vertx.core.impl.BlockedThreadChecker] (vertx-blocked-thread-checker) Thread Thread[vert.x-eventloop-thread-1,5,main] has been blocked for 2374 ms, time limit is 2000 ms: io.vertx.core.VertxException: Thread blocked
Not sure what is going wrong here but it worked if started as [start-dev]

[start-dev]
/opt/keycloak/bin/kc.sh show-config
Current Mode: development
Current Configuration:
	kc.cache =  local (PersistedConfigSource)
	kc.config.built =  true (SysPropConfigSource)
	kc.db =  dev-file (PersistedConfigSource)
	kc.health-enabled =  false (PersistedConfigSource)
	kc.hostname-strict =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.hostname-strict-https =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.http-enabled =  true (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.http-relative-path =  / (PersistedConfigSource)
	kc.log-console-output =  default (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.log-file =  ${kc.home.dir:default}${file.separator}data${file.separator}log${file.separator}keycloak.log (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.metrics-enabled =  false (PersistedConfigSource)
	kc.proxy =  edge (KcEnvVarConfigSource)
	kc.spi-map-storage-concurrenthashmap-dir =  ${kc.home.dir:default}${file.separator}data${file.separator}chm (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.spi-map-storage-concurrenthashmap-key-type-authz-resource-servers =  string (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.spi-map-storage-concurrenthashmap-key-type-realms =  string (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.spi-map-storage-concurrenthashmap-key-type-single-use-objects =  string (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.spi-theme-cache-templates =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.spi-theme-cache-themes =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.spi-theme-static-max-age =  -1 (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-19.0.2.jar!/META-INF/keycloak.conf])
	kc.version =  19.0.2 (SysPropConfigSource)

mysql> use keycloak;
Database changed
mysql> show tables;
Empty set (0.01 sec)

_On a side note when a container was created uding jboss/keycloak:16.1.1 image, I was able to see a bunch of tables but it failed constantly giving a timeout issue at a later stage. It is different issue but with the same DB configuration my minikube container is able to talk to local db_

### Version

19.0.2

### Expected behavior

Data should be persisted into MySQL. If pod is deleted and recreated I should be able to retain my realm, configuraiton and users.

**Comment by stianst** [comment]
I've seen the `io.vertx.core.VertxException: Thread blocked` issue on CI a few times, so there may be some sort of race-condition here on startup.

@enfochannel does this happen every-time? Have you tried restarting the container?

**Comment by enfochannel** [comment]
The only way I can resume the service is by killing the pod, or worst stopping and starting minikube, or worst deleting the deployment. In any case, since data is not persisted then it has to restart the whole thing. I can give it a try without DB-specific environment parameters but that is not the goal :)

**Comment by vmuzikar** [comment]
@enfochannel Hello, thanks for the report. Could you please try it with Docker as the minikube driver? If I recall correctly, there were several (network related) bugs in the HyperKit driver.

**Comment by enfochannel** [comment]
The Docker driver is not reliable for ingress and also applications running on minikube cannot communicate with external MySQL. I tried  KC_DB_URL "127.0.0.1" and also IP from minikube internal host (host.minikube.internal) none of the IPs worked. I changed the configuration as per this [link](https://www.keycloak.org/server/db) but no luck either.

```
minikube start \
--cpus=2 \
--memory=1990 \
--driver=docker \  
--vm=true
```
```
minikube ssh
docker@minikube:~$ ping host.minikube.internal
PING host.minikube.internal (192.168.65.2) 56(84) bytes of data.
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=1 ttl=37 time=4.77 ms
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=2 ttl=37 time=0.829 ms
64 bytes from host.minikube.internal (192.168.65.2): icmp_seq=3 ttl=37 time=1.56 ms
^C
--- host.minikube.internal ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2004ms
rtt min/avg/max/mdev = 0.829/2.384/4.765/1.709 ms

---

docker@minikube:~$ ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.148 ms
64 bytes from 127.0.0.1: icmp_seq=2 ttl=64 time=0.068 ms
64 bytes from 127.0.0.1: icmp_seq=3 ttl=64 time=0.036 ms
64 bytes from 127.0.0.1: icmp_seq=4 ttl=64 time=0.071 ms
^C
--- 127.0.0.1 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3103ms
rtt min/avg/max/mdev = 0.036/0.080/0.148/0.041 ms
```
```
docker@minikube:~$ nc 192.168.65.2 3306
J
8.0.31n;Uy{
           ?????>Xu;iukkQcaching_sha2_password^C

Assuming connection is successful
```
```
mysql -h 192.168.49.2 -u root -p
Enter password: 
ERROR 2003 (HY000): Can't connect to MySQL server on '192.168.49.2:3306' (60)
```
So the only option to connect to db is using 127.0.0.1. Also as shown above minikube can ping 127.0.0.1 but cannot connect to port 3306.
```
docker@minikube:~$ nc 127.0.0.1 3306
docker@minikube:~$ 
```

```
        image: quay.io/keycloak/keycloak:19.0.2
        args: ["start-dev"]
        env:
        - name: KEYCLOAK_ADMIN
          value: "admin"
        - name: KEYCLOAK_ADMIN_PASSWORD
          value: "password1234"
        - name: KC_PROXY
          value: "edge"
        - name: KC_DB
          value: mysql
        - name: KC_DB_URL
          value: "127.0.0.1"
        - name: KC_DB_URL_PORT
          value: "3306"
        - name: KC_DB_URL_DATABASE
          value: "keycloak"
        - name: KC_DB_USERNAME
          value: "root"
        - name: KC_DB_PASSWORD
          value: "password1234"
        - name: JDBC_PARAMS
          value: "serverTimezone=UTC"
```

Error:
```
2022-10-21 21:22:00,912 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': Connector/J cannot handle a connection string '127.0.0.1'.
2022-10-21 21:22:00,919 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: Connector/J cannot handle a connection string '127.0.0.1'.
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:131)
	at com.mysql.cj.jdbc.MysqlDataSource.getConnection(MysqlDataSource.java:128)
	at com.mysql.cj.jdbc.MysqlDataSource.getConnection(MysqlDataSource.java:113)
	at com.mysql.cj.jdbc.MysqlXADataSource.getXAConnection(MysqlXADataSource.java:52)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:216)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:513)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:494)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:75)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

2022-10-21 21:27:24,816 INFO  [org.keycloak.common.crypto.CryptoIntegration] (main) Detected crypto provider: org.keycloak.crypto.def.DefaultCryptoProvider
2022-10-21 21:27:26,484 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': Connector/J cannot handle a connection string '192.168.65.2'.
2022-10-21 21:27:26,491 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: Connector/J cannot handle a connection string '192.168.65.2'.

Tried even GCP MySQL:

2022-10-21 21:30:45,400 INFO  [org.keycloak.common.crypto.CryptoIntegration] (main) Detected crypto provider: org.keycloak.crypto.def.DefaultCryptoProvider
2022-10-21 21:30:46,945 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': Connector/J cannot handle a connection string '34.79.75.144'.
2022-10-21 21:30:46,955 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: Connector/J cannot handle a connection string '34.79.75.144'.
```
Basically, nothing seems working.

**Comment by enfochannel** [comment]
Ok, finally the issue is solved - Once again thanks for all the help.

I referred to: docker-compose-examples do not work w/ KC 17+ #388

Basically, the parameters were already amended as mentioned above but the URL was missing the full JDBC path

```
 - name: KC_DB_URL
          value: "jdbc:mysql://host.docker.internal:3306/keycloak"
```
So other environmental parameters are redundant?:

```
- name: KC_DB_URL_PORT
          value: "3306"
        - name: KC_DB_URL_DATABASE
          value: "keycloak"
```
Also, since never imagined that the hostname would work like host.docker.internal, but it seems the only way the container can reach out to an externally hosted database. It is worth giving a try with cloud SQL.

So the final manifest is like this:

```
apiVersion: v1
kind: Service
metadata:
  name: keycloak
  labels:
    app: keycloak
spec:
  ports:
  - name: http
    port: 8080
    targetPort: 8080
  selector:
    app: keycloak
  type: LoadBalancer
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: keycloak
  labels:
    app: keycloak
spec:
  replicas: 1
  selector:
    matchLabels:
      app: keycloak
  template:
    metadata:
      labels:
        app: keycloak
    spec:
      containers:
      - name: keycloak
        #image: jboss/keycloak:16.1.1
        image: quay.io/keycloak/keycloak:19.0.2
        args: ["start-dev"]
        env:
        - name: KEYCLOAK_ADMIN
          value: "admin"
        - name: KEYCLOAK_ADMIN_PASSWORD
          value: "xxxxx"
        - name: KC_PROXY
          value: "edge"
        - name: KC_DB
          value: mysql
        - name: KC_DB_URL
          value: "jdbc:mysql://host.docker.internal:3306/keycloak"
        - name: KC_DB_URL_PORT
          value: "3306"
        - name: KC_DB_URL_DATABASE
          value: "keycloak"
        - name: KC_DB_USERNAME
          value: "root"
        - name: KC_DB_PASSWORD
          value: "xxxxx"
        - name: JDBC_PARAMS
          value: "serverTimezone=UTC"
        #- name: KC_HOSTNAME
        #  value: keycloak.192.168.49.2.io
        ports:
        - name: http
          containerPort: 8080
        readinessProbe:
          httpGet:
            path: /realms/master
            port: 8080
      enableServiceLinks: false
```
For now, I am keeping those additional environmental parameters as is.

DB is now updated with values:
```
mysql> select count(name) from realm;
+-------------+
| count(name) |
+-------------+
|           2 |
+-------------+
1 row in set (0.00 sec)
```
Thanks once again ^^
